dpgen2/exploration/report/report_adaptive_lower.py

```diff
@@ -0,0 +1,330 @@
+        accur = set()
+        coll_f = []
+        coll_v = []
+        for ii in range(nframes):
+            if md_f[ii] > self.level_f_hi or md_v[ii] > self.level_v_hi:
+                failed.append((tt, ii))
+            else:
+                coll_f.append([md_f[ii], tt, ii])
+                coll_v.append([md_v[ii], tt, ii])
+                # now accur takes all non-failed frames,
+                # will be substracted by candidate later
+                accur.add((tt, ii))
+        return nframes, accur, failed, coll_f, coll_v
+
+    def _sequence_conv(
+            self,
+            seq,
+    )->bool:
+        if len(seq) <= 1:
+            return False
+        conv = [ abs(seq[ii-1] - seq[ii]) < self.conv_tolerance 
+                 for ii in range(1,len(seq)) ]
+        return all(conv)
+
+    def converged(
+            self,
+            reports,
+    )->bool:
+        if 1 + len(reports) < self.n_checked_steps:
+            return False
+        else:
+            all_level_f = [ii.level_f_lo for ii in reports] + [self.level_f_lo]
+            all_level_f = all_level_f[-self.n_checked_steps:]
+            conv = self._sequence_conv(all_level_f)
+            if self.has_virial:
+                all_level_v = [ii.level_v_lo for ii in reports] + [self.level_v_lo]
+                all_level_v = all_level_v[-self.n_checked_steps:]
+                conv = conv and self._sequence_conv(all_level_v)
+            return conv            
+
+    def failed_ratio(
+            self,
+            tag = None,
+    ):
+        return float(len(self.failed)) / float(self.nframes)
+
+    def accurate_ratio(
+            self,
+            tag = None,
+    ):
+        return float(len(self.accur)) / float(self.nframes)
+
+    def candidate_ratio(
+            self,
+            tag = None,
+    ):
+        return float(len(self.candi)) / float(self.nframes)
+
+    def get_candidate_ids(
+            self,
+            max_nframes : Optional[int] = None,
+    ) -> List[List[int]]:
+        ntraj = self.ntraj
+        id_cand = self._get_candidates(max_nframes)
+        id_cand_list = [[] for ii in range(ntraj)]
+        for ii in id_cand:
+            id_cand_list[ii[0]].append(ii[1])
+        return id_cand_list
+
+    def _get_candidates(
+            self,
+            max_nframes : Optional[int] = None,
+    )->List[Tuple[int,int]]:
+        """
+        Get candidates. If number of candidates is larger than `max_nframes`, 
+        then randomly pick `max_nframes` frames from the candidates. 
+
+        Parameters
+        ----------
+        max_nframes    int
+                The maximal number of frames of candidates.
+
+        Returns
+        -------
+        cand_frames   List[Tuple[int,int]]
+                Candidate frames. A list of tuples: [(traj_idx, frame_idx), ...]
+        """
+        self.candi_picked = [(ii[0], ii[1]) for ii in self.candi]
+        if max_nframes and max_nframes < len(self.candi_picked):
```

Comment:
I am not sure if `max_nframes = 0` is accepted. But to handle this situation, `if max_nframes and ...` should be `if max_nframes is not None and ...`.

Reply:
Right, fixed, thanks!